dashboard/src/components/AppView/testdata/test-k8s-manifest.tsx

```diff
@@ -0,0 +1,45 @@
+const manifest = `
```

Comment:
I am not a big fan of static files for unit testing because:

 - They are not directly related to a specific unit test. What parts are you interested in? where are they used?
 - They are not reusable. You cannot extend/modify a file since it may break an existing unit test. This lead to the growth of these kind of files which become very difficult to manage.

I'd rather move this kind of manifest to a factory method (or just static vars) that returns only the piece of manifest you are interested in.

Reply:
I do not think that I fully understand the concern so I'd like to know more.

> They are not directly related to a specific unit test. What parts are you interested in? where are they used?

Right, but they are related to a specific component and hence is inside component/testdata. Similarly you can find it in gocode, it might not be idiomatic here though. 

The reason I use a plain manifest is to test the convoluted componentWillReceiveProps which parses the yaml, splits resources by the separator and puts them in the different buckets.

This manifest also contains invalid k8s resources to simulate a release with wrong data (like one with just a comment). So I am interested in the whole file not just parts of it.

> They are not reusable. You cannot extend/modify a file since it may break an existing unit test. This lead to the growth of these kind of files which become very difficult to manage.

I agree, but that's not a problem we don't have yet. I am not even sure if future tests of that component will require the manifest at all, since they might be testing something else that is not related to the list of k8s resources of a chart. It also seemed to me like a good compromise in simplicity.

> I'd rather move this kind of manifest to a factory method (or just static vars) that returns only the piece of manifest you are interested in.

I want to know more about this. Could you add an example. Please note that I am interested in the whole file because the whole resources extraction is coupled.

```js
let manifest: IResource[] = yaml.safeLoadAll(newApp.manifest);
    // Filter out elements in the manifest that does not comply
    // with { kind: foo }
    manifest = manifest.filter(r => r && r.kind);

    const watchedKinds = ["Deployment", "Service"];
    const otherResources = manifest
      .filter(d => watchedKinds.indexOf(d.kind) < 0)
      .reduce((acc, r) => {
        // TODO: skip list resource for now
        if (r.kind === "List") {
          return acc;
        }
        acc[`${r.kind}/${r.metadata.name}`] = r;
        return acc;
      }, new Map<string, IResource>());
    this.setState({ otherResources });

    const deployments = manifest.filter(d => d.kind === "Deployment");
    const services = manifest.filter(d => d.kind === "Service");
```

Reply:
> This manifest also contains invalid k8s resources to simulate a release with wrong data (like one with just a comment). So I am interested in the whole file not just parts of it.

That's mostly what I would like to avoid. You are using the same manifest to test several things. For example to test that is properly extracting a deployment and a service (for the sockets thing) but you are using the same thing for testing that it ignores invalid objects. That's difficult to discover for others (I for example didn't notice the invalid object part).

I'd split that so you have different self-contained test. If you are testing that the code is ignoring invalid objects you clearly set a manifest as simple as possible that let you test that feature.

> I want to know more about this. Could you add an example.

It doesn't have to be very smart. Maybe a simple function that return a YAML text from a object. For example for the sockets test:

```
it("sets a list of web sockets for its deployments and services", () =>
   manifest = generateManifest([
      {apiVersion: "extensions/v1beta1", kind: "Deployment", metadata: { ... }, spec: { ... }},
      {apiVersion: "v1", kind: "Service", metadata: { ... }, spec: { ... }},
   ]);
   const validProps = getValidProps(manifest);
   (the rest of the test...)
```

Reply:
I agree that it would be nicer to see more self-defined tests here rather than just a comment that spurious bogus resources aren't in the state. The generateManifest helper is cool, but how would that be used to add the empty manifests?

Reply:
I think we are talking about two different things here:

1 - Use a single purpose file for multiple test cases

The generator idea makes sense but the main reason for not crafting my own manifests was to test that the parser actually parses different resources separated by the `---` separator which is what we get from the API response.

I agree that it is not clear that in that manifest there are bogus data, so what I can do is to separate this into multiple files, would that work?

2 - The tests cases do too much and it is not clear what

I agree, so I'll split it.

Would these two action items work for you both?

Reply:
@migmartri whilst you could use a single manifest file, doesn't that make it more difficult to maintain it? For example, you can test that empty YAML is ignored with a smaller manifest that will be easier to maintain and understand what it is used for:

```
---
kind: Service
metadata:
  name: manifest-with-empty-files
...
---
# I'm empty with a comment
---
---
```

Reply:
My main concern is to have tests that depends from other files when it's not very clear how. That's what I find difficult to maintain.

> The generator idea makes sense but the main reason for not crafting my own manifests was to test that the parser actually parses different resources separated by the --- separator which is what we get from the API response.

You can still add the `---` separator in the generator function (sorry that I didn't make it clear). For example (untested code):

```
function generateManifest(items) {
 let manifest = "";
 items.forEach(i => {
   manifest += "---\n" + yaml.safeDump(i);
 }
 return manifest;
```

> I agree that it is not clear that in that manifest there are bogus data, so what I can do is to separate this into multiple files, would that work?

I still don't like that very much because it's difficult to relate external files to unit tests.

> 2 - The tests cases do too much and it is not clear what
> I agree, so I'll split it.

That's true but that's not my main concern.

Reply:
Thanks @andresmgot @prydonius 

I have update the test cases by using the generator plus splitting the test cases.

Thanks for the help!